File: code_eval/java/easy/Challenge03.java
package easy;

/**
 * Challenge 03 : Largest Palindrome prime < 1000
 * Created by cschalk on 3/21/16.
 */
public class Challenge03 {

    public static void main(String[] args) {
        for (int i = 1000; i > 0; i--) {
            if (isPrime(i) && isPalindrome(i)) {
                System.out.println(i);
                System.exit(0);
            }
        }
    }

    private static boolean isPrime(int in) {
        switch (in) {
            case 1  : return false;
            case 2  : return true;
            default :
                // get rid of evens
                if (in % 2 == 0) return false;

                // efficiently loop through rest
                for (int i = 3; i < Math.sqrt(in) + 1; i++) {
                    if (in % i == 0) return false;
                }
        }
        return true;
    }

    private static boolean isPalindrome(int in) {
        String strIn = "" + in;
        for (int i = 0, j = strIn.length() - 1; i < j ; i++, j-- ) {
            if (strIn.charAt(i) != strIn.charAt(j)) {
                return false;
            }
        }
        return true;
    }
}

File: code_eval/cpp/easy/challenge_03.cpp
// Created by cschalk on 3/21/16.
// Challenge 03 : Largest prime palindrome less than 1000
#include <iostream>
#include <math.h>
using namespace std;

bool isPrime(int test) {
    switch(test) {
        case 1  : return false;
        case 2  : return true;
        default : {
            // get rid of evens first
            if (test % 2 == 0) return false;

            // test only odd divisors in the loop
            auto root = sqrt(test) + 1;
            for (auto mod = 3; mod < root; mod += 2) {
                if (test % mod == 0) return false;
            }
            return true;
        }
    }
}

bool isPalindrome(int input) {
    std::string test = std::to_string(input);
    for (unsigned long i = 0, j = test.length() - 1 ; i < j ; i++, j-- ) {
        if (test[i] != test[j]) {
            return false;
        }
    }
    return true;
}

int main(int argc, char ** argv) {
    for (int i = 1000; i > 0 ; i--) {
        if (isPrime(i)) {
            if (isPalindrome(i)) {
                std::cout << i << std::endl;
		return 0;
            }
        }
    }
    return 1;
}
